module/zfs/dsl_bookmark.c

```diff
@@ -116,57 +120,177 @@ dsl_bookmark_lookup(dsl_pool_t *dp, const char *fullname,
 	return (error);
 }
 
-typedef struct dsl_bookmark_create_redacted_arg {
-	const char	*dbcra_bmark;
-	const char	*dbcra_snap;
-	redaction_list_t **dbcra_rl;
-	uint64_t	dbcra_numsnaps;
-	uint64_t	*dbcra_snaps;
-	void		*dbcra_tag;
-} dsl_bookmark_create_redacted_arg_t;
+/*
+ * Check that the given nvlist corresponds to the following schema:
+ *  { newbookmark -> source, ... }
+ * where
+ * - a newbookmark is a full dataset path of a bookmark (bookmark_namecheck)
+ * - a source is a full path of a snaphot or bookmark
+ *   ({bookmark,snapshot}_namecheck)
+ * - all newbookmarks are in the same pool
+ * - all newbookmarks have unique names
+ * - FIXME: need to check that newbookmark is never used as source in any other?
+ *
+ * Note that this does not check for existence of the pool or source bookmarks.
+ *
+ * Returns 0 if the nvlist adheres to above schema.
+ * Returns -1 if it doesn't.
+ */
+int
+dsl_bookmark_create_nvl_validate(nvlist_t *bmarks)
+{
+	char *first, *source;
+	size_t first_len;
+
+	first = NULL;
+	for (nvpair_t *pair = nvlist_next_nvpair(bmarks, NULL);
+	    pair != NULL; pair = nvlist_next_nvpair(bmarks, pair)) {
+		int is_bmark, is_snap;
+
+		char *bmark = nvpair_name(pair);
+
+		if (bookmark_namecheck(bmark, NULL, NULL) != 0)
+			return (-1);
+
+		/* structure check: values must be snapshots XOR bookmarks */
+		if (nvpair_value_string(pair, &source) != 0)
+			return (-1);
+		is_bmark = bookmark_namecheck(source, NULL, NULL) == 0;
+		is_snap = snapshot_namecheck(source, NULL, NULL) == 0;
+		if (!is_bmark && !is_snap) {
+			return (-1);
+		}
+
+		/* same pool check */
+		if (first == NULL) {
+			char *cp = strpbrk(bmark, "/#");
+			if (cp == NULL)
+				return (-1);
+			first = bmark;
+			first_len = cp - bmark;
+		}
+
+		if (strncmp(first, bmark, first_len) != 0)
+			return (-1);
+
+		if (*(bmark + first_len) != '/')
+			return (-1);
 
-typedef struct dsl_bookmark_create_arg {
-	nvlist_t *dbca_bmarks;
-	nvlist_t *dbca_errors;
-} dsl_bookmark_create_arg_t;
+		/* unique newbookmark names; todo: O(n^2) */
+		for (nvpair_t *pair2 = nvlist_next_nvpair(bmarks, pair);
+		    pair2 != NULL; pair2 = nvlist_next_nvpair(bmarks, pair2)) {
+			if (strcmp(nvpair_name(pair), nvpair_name(pair2)) == 0)
+				return (-1);
+		}
```

Comment:
Why is this needed?  We already do it in zfs_ioc_bookmark(), and the ZCP code only has one entry in the nvlist.

Reply:
I think it makes sense to have _one_ function that validates the schema of the bookmark_create nvlist, and have that function be in the common code path for both ioctl and zcp (both ioctl and zcp call `dsl_bookmark_create_check`, which in turn calls this function).

I propose to remove the uniqueness check in the ioctl handler.
It shouldn't make a noticable performance difference for the zcp use case with a single-pair nvlist.